tlatools/org.lamport.tlatools/test/pcal/parser/PcalParserTranslator.java

```diff
@@ -0,0 +1,184 @@
+package pcal.parser;
+
+import java.util.Arrays;
+import java.util.Vector;
+
+import util.AstNode;
+import util.AstNode.Kind;
+
+import pcal.AST;
+import pcal.PcalCharReader;
+import pcal.TLAtoPCalMapping;
+import pcal.TLAExpr;
+import pcal.TLAToken;
+import pcal.ParseAlgorithm;
+import pcal.PcalParams;
+import pcal.exception.ParseAlgorithmException;
+
+import org.junit.Assert;
+
+/**
+ * Utilities for parsing PlusCal and converting it to a form compatible with
+ * the unified TLA+ syntax parser test corpus.
+ */
+public class PcalParserTranslator {
+	
+	/**
+	 * Simple data class holding info about the start of a PlusCal block.
+	 */
+	private static class PcalStart {
+		/**
+		 * The line on which the PlusCal block starts.
+		 */
+		int line = 0;
+		
+		/**
+		 * The column in which the PlusCal block starts, directly after the
+		 * --algorithm or --fair algorithm tokens.
+		 */
+		int column = 0;
+		
+		/**
+		 * Whether the PlusCal block is a fair algorithm.
+		 */
+		boolean isFair = false;
+		
+		/**
+		 * Finds the start of the PlusCal block.
+		 * 
+		 * @param input Text containing PlusCal block, split into lines.
+		 */
+		public PcalStart(Vector<String> input) {
+			final String fairStart = "--fair";
+			final String unfairStart = "--algorithm";
```

Comment:
These constants are already defined in `PcalParams`. That file also lists
```java
public static final String BeginFairAlg2 = "algorithm" ;
```
Is that a possible begin token that this class needs to handle?

Reply:
Good catch, will switch to using those.

The weird expected behavior of the `ParseAlgorithm.getAlgorithm` method is that, if we are parsing a fair algorithm (specified with a boolean parameter), it expects to find the `algorithm` keyword. If we are not parsing a fair algorithm, it expects to start parsing the PlusCal code block directly.

Reply:
Please consider adding a corresponding comment to `ParseAlgorithm.getAlgorithm`.

Reply:
It is currently documented (in non-javadoc style):

https://github.com/tlaplus/tlaplus/blob/a15241e40c56284e36c451a9b65159ab38990aa1/tlatools/org.lamport.tlatools/src/pcal/ParseAlgorithm.java#L308-L311